### DISCUSSION TITLE
What is ReOptimizeTNLP() function in IPOPT C++ interface and how to use it?

### DISCUSSION THREAD

**Discussion by yanhaoy** [coin-or/Ipopt#499] [Q&A] [answered]
I am using IPOPT to solve my MPC and try to speed it up. I have turned on the warm start and got a significant improvement. However, I found that there is a function handle called ReOptimizeTNLP(), but it is not mentioned in the C++ interface documentation. I found that some third-party code uses this function starting from the second optimization. I checked the code a bit and found that this method reuses TNLPAdapter. In actual use, it seems that using ReOptimizeTNLP() will make the number of iterations after the warm start more stable and keep smaller. What is the difference between ReOptimizeTNLP() and OptimizeTNLP()? How is it different from directly enabling the warm start? When or under what conditions can it be used? Thank you very much!

**Comment by svigerske** [comment, marked as answer]
The [docu of ReOptimizeTNLP](https://coin-or.github.io/Ipopt/classIpopt_1_1IpoptApplication.html#a24de5d491b570b01e36a329c5a3c9471) says a bit more, though I don't think that the structure must be the same. ReOptimizeTNLP still does a lot of initialization, which would deal with changes in problem structure.
Option [warm_start_same_structure](https://coin-or.github.io/Ipopt/OPTIONS.html#OPT_warm_start_same_structure) needs to be set additionally if more initializations should be skipped.

So, you can use `ReOptimizeTNLP` whenever you solve with the same `TNLP` object that you used for solving before. This will skip reconstruction of the `TNLPAdapter` that translates between the `TNLP` and the NLP representation that Ipopt uses internally. It will also skip reconstruction of the `IpoptAlgorithm` object, so if you change Ipopt options that would switch components of the Ipopt algorithm (like using a different linear solver), then you should not use `ReOptimizeTNLP`.

If in addition the problem structure does not change, that is, the Jacobian and Hessian structure does not change, variable bounds do not appear or disappear, constraint sides do not appear or disappear, and constraints do not change between equality and ranged constraints, then setting option `warm_start_same_structure` could save a bit of time. This also prevents that the scaling is recomputed (see also #481).

`warm_start_same_structure` doesn't have anything to do with `warm_start_init_point`. So calling `ReOptimizeTNLP` or using `warm_start_same_structure` doesn't actually do warmstarts in the sense that numerical values from the previous solve are reused, I believe (the scaling parameters are an exception; one should check carefully if there is more). My understanding is that they just skip re-initialization of data structures that would otherwise turn out completely identical.

**Reply by yanhaoy** [reply]
Hi Stefan, thank you very much for your answer.

If I understand correctly, there is no overlap between `ReOptimizeTNLP`, `warm_start_same_structure` and `warm_start_init_point`. So if the corresponding conditions are met, I can turn them on, right?

Regarding the scaling parameter, I saw somewhere that if I want to use warm start, I should set `nlp_scaling_method` to `none`. Otherwise, it will scale the problem according to the initial gradient, and the previous solution is meaningless. Is this correct?

Thank you!

**Reply by svigerske** [reply]
Yes, there is not much of an overlap. I don't think that it makes sense to use `OptimizeTNLP` with `warm_start_same_structure=yes`. But `warm_start_init_point` makes sense independent of the other two.

The "previous solution" is for the unscaled problem, so it doesn't become meaningless if the scaling changes, I believe. 
I think that the [suggestion to turn off scaling](https://list.coin-or.org/pipermail/ipopt/2016-May/004221.html) was motivated by trying to ensure that when doing a warmstart on the exact same problem, Ipopt should accept the given warmstart immediately and not do any iterations.    
The discussion #481 suggests that it should be easier to do the scaling independent of the starting point, because when you solve a sequence of very similar problems and try to do warmstarts, you might want to keep the scaling constant. Turning off scaling is obviously one way to achieve that, but I wouldn't be sure that this is usually better than having the scaling change from one solve to another. Providing scaling factors by yourself (`nlp_scaling_method=user_scaling`) is another way to keep the scaling constant. Finally, if you can set `warm_start_same_structure=yes`, then scaling factors from the previous solve are also preserved.
